crates/spk-schema/src/v0/package_spec.rs

```diff
@@ -0,0 +1,668 @@
+// Copyright (c) Contributors to the SPK project.
+// SPDX-License-Identifier: Apache-2.0
+// https://github.com/spkenv/spk
+
+use std::borrow::Cow;
+use std::collections::BTreeSet;
+use std::convert::TryInto;
+use std::str::FromStr;
+
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use spk_schema_foundation::IsDefault;
+use spk_schema_foundation::ident::{AsVersionIdent, BuildIdent};
+use spk_schema_foundation::option_map::{OptFilter, Stringified};
+use spk_schema_foundation::version::{
+    BuildIdProblem,
+    CommaSeparated,
+    ComponentsMissingProblem,
+    IncompatibleReason,
+    PackageNameProblem,
+    VarOptionProblem,
+};
+
+use super::TestSpec;
+use crate::build_spec::UncheckedBuildSpec;
+use crate::foundation::ident_build::Build;
+use crate::foundation::ident_component::Component;
+use crate::foundation::name::PkgName;
+use crate::foundation::option_map::OptionMap;
+use crate::foundation::spec_ops::prelude::*;
+use crate::foundation::version::{Compat, CompatRule, Compatibility, Version};
+use crate::foundation::version_range::Ranged;
+use crate::ident::{
+    PkgRequest,
+    PreReleasePolicy,
+    Request,
+    RequestedBy,
+    Satisfy,
+    VarRequest,
+    is_false,
+};
+use crate::metadata::Meta;
+use crate::option::VarOpt;
+use crate::v0::EmbeddedPackageSpec;
+use crate::{
+    BuildSpec,
+    ComponentSpec,
+    ComponentSpecList,
+    Components,
+    Deprecate,
+    DeprecateMut,
+    EmbeddedPackagesList,
+    EnvOp,
+    Inheritance,
+    InstallSpec,
+    LocalSource,
+    Opt,
+    Package,
+    PackageMut,
+    RequirementsList,
+    Result,
+    RuntimeEnvironment,
+    SourceSpec,
+    ValidationSpec,
+};
+
+#[cfg(test)]
+#[path = "./package_spec_test.rs"]
+mod package_spec_test;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Ord, PartialOrd, Serialize)]
+pub struct PackageSpec {
```

Comment:
just noting that some of these "top-level" structs don't have any descriptions, might be a useful addition if you are feeling inclined